src/main/java/com/google/devtools/build/lib/remote/blobstore/http/HttpDownloadHandler.java

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package com.google.devtools.build.lib.remote.blobstore.http;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.auth.Credentials;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.HttpContent;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.internal.StringUtil;
+import java.io.IOException;
+import java.io.OutputStream;
+
+class HttpDownloadHandler extends AbstractHttpHandler<HttpObject> {
+
+  private OutputStream out;
+  private boolean keepAlive = HttpVersion.HTTP_1_1.isKeepAliveDefault();
+
+  public HttpDownloadHandler(Credentials credentials) {
+    super(credentials);
+  }
+
+  @Override
+  protected void channelRead0(ChannelHandlerContext ctx, HttpObject msg) throws Exception {
+    checkState(userPromise != null, "response before request");
+    if (msg instanceof HttpResponse) {
+      HttpResponse response = (HttpResponse) msg;
+      keepAlive = HttpUtil.isKeepAlive((HttpResponse) msg);
+      if (!response.status().equals(HttpResponseStatus.OK)) {
+        failAndReset(
+            new HttpException(
+                response.status(), "Download failed with Status: " + response.status(), null),
+            ctx);
+      }
+    } else if (msg instanceof HttpContent) {
+      ByteBuf content = ((HttpContent) msg).content();
+      content.readBytes(out, content.readableBytes());
```

Comment:
OutputStream could block?

Reply:
Yes. Upload and download handler should be run in their own thread pool or disk i/o should be implemented async. I didn't want to introduce yet another threadpool thinking that this will be "fast enough". Not sure?